Query Refactoring: validate GeoShapeQueryBuilder strategy and relation setting

cbuescher opened the pull request:
Before the refactoring we didn't check any invalid settings for strategy and relation in the GeoShapeQueryBuilder. However, using SpatialStrategy.TERM and ShapeRelation.INTERSECTS together is invalid and we tried to protect against that in the validate() method.
This PR moves these checks to setter for strategy and relation and adds tests for the new behaviour.

Relates to #10217

cbuescher commented:
I'm not super happy with this solution since it seems to me that relation() and strategy(), although related by this one illegal pairing, should not simply be combined in one setter like I did for demonstration purposes in this PR. However, I don't see many other options besides delaying the validation to the shard (in `toQuery`) at this point. Happy to get feedback on this.

cbuescher commented:
Thinking about this a bit longer, I would probably got for separate setters again, since the two parameters are not really tightly related. Many tests just set `relation` without even specifying the optional `stragegy`. So what if we:
* on setting `relation()`, check that either there was no strategy set before, otherwise implicitely set it to INTERSECTS.
* on setting strategy(), check that the chosen strategy is compatible with the currently set relation, otherwise throw IAE
Any opinions about this solution?

javanna commented on core/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java:
```diff
@@ -183,6 +183,10 @@ public String indexedShapeType() {
      * @return this
      */
     public GeoShapeQueryBuilder strategy(SpatialStrategy strategy) {
+        if (strategy != null && strategy == SpatialStrategy.TERM && relation != ShapeRelation.INTERSECTS) {
```
does it make sense to check for null beforehand and throw exception? we don't want anybody to call this method and explicitly pass in `null` I guess.

cbuescher replied:
The default for `strategy` currently is to be unset (`null` that is), in which case the field types default strategy is going to be used later when building the lucene query. That's why I thought it's okay if the setter is used with `null` value as well, although it would be a no-op. Might be even useful if `strategy` was set to something else and needs to be changed back to the default.

javanna commented:
I left a small comment. Regarding your question, I am on the fence...it is not perfect either way to me but I don't know how to do it better, I wouldn't delay this validation to the data node. I think you can leave what you have now, that's fine.

cbuescher commented:
@javanna left a comment, I think I should leave that one without a null check for the reasons given above. Also squashed and rebased, should I wait another review round or do you think this is ready to go?

javanna commented:
@s1monw @colings86 what do you think about the validation in the setter depending on the current state of the object? I am ok with it...

colings86 commented:
@javanna @cbuescher I think the logic needs to be a bit different from what is described above though. I think it should be:
* on setting relation(), check that either there was no strategy set before, if there is then:
    * if the strategy is RECURSIVE set the relation
    * if the strategy is TERM then:
        * if the relation in the parameter is INTERSECTS set the relation
        * if the relation is anything else throw an IllegalArgumentException
* on setting strategy(), check that the chosen strategy is compatible with the currently set relation, otherwise throw an IllegalArgumentException (basically throw the exception if the strategy is TERM and the relation is NOT EQUAL to INTERSECTS)

cbuescher commented:
@colings86 you're right, the description I added previously was off, but what you described is basically whats done in the code in this PR. Want to recheck and tell me if this looks okay to you?

colings86 commented on core/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java:
```diff
@@ -183,6 +183,10 @@ public String indexedShapeType() {
      */
     public GeoShapeQueryBuilder strategy(SpatialStrategy strategy) {
+        if (strategy != null && strategy == SpatialStrategy.TERM && relation != ShapeRelation.INTERSECTS) {
+            throw new IllegalArgumentException("current strategy [" + strategy.getStrategyName() + "] only supports relation ["
```
I would remove the word current here as its the strategy the user is trying to set that is the issue not the one already set

colings86 commented:
@cbuescher yep you are right, the code follows what I detailed above. I left a minor comment on the exception wording though

cbuescher commented:
Rebased and addressed the exception wording @colings86 mentioned. Anything else to add here?
